# Example 2: Annuity with DAV Annuity Table

In the current example we build on the previous one and demonstrate two additional things:

  * using the DAV2004R table as the base mortality assumption
  * starting the projector from code
 
 
 The DAV2004R table can be used in a variety of ways for different use cases and ```pyprotolinc``` provides
 some convenience code to use some of those with little effort.
 
 To follow along change into the directory *examples/02_annuities_in_payment*.

### Download the DAV Tables

It's not so clear why the DAV tables aren't easily available for downloading from official sources. Luckily, the maintainer of the R-Project *MortalityTables* (cf. https://www.rdocumentation.org/packages/MortalityTables/versions/2.0.3) provides the required files. If you are willing to get them (the legal situation around this is not clear to the author of this tutorial) open up a command prompt and change into the base directory of this example and run:
```
pyprotolinc download_dav_tables
```
**or** run the folling code from a script:

In [1]:
from pyprotolinc.utils import download_dav_tables
download_dav_tables()

Now a new folder called *tables* should have been created in  the working folder with two subfolders. In the present case the one of interest is this:

In [2]:
# use when not on Windows "!ls"
!dir /B tables\Germany_Annuities_DAV2004R

Germany_Annuities_DAV2004R.csv
Germany_Annuities_DAV2004R_AV.csv
Germany_Annuities_DAV2004R_AVBase.csv
Germany_Annuities_DAV2004R_Select.csv
Germany_Annuities_DAV2004R_Trends.csv


*PyProtolinc* provides an interface class to these files. On instantiation pass the base folder from which to load the files as an argument.

In [3]:
from pyprotolinc.assumptions.dav2004r import DAV2004R
dav2004r = DAV2004R(base_directory="tables/Germany_Annuities_DAV2004R", trend_t1=10, trend_t2=25)

The second and third argument above control the trend factor extrapolation between the start and target trends.


As an aside it is quickly demonstrated here how one can work with this *raw* object (i.e. outside the context of a projection run). In this case one first has to create a *rates provider* object. Specifically, in this case we parametrize the *aggregate best estimate rates*, i.e. the aggregate (non-select) version using no additional prudency loadings ("2. Ordnung").

In [4]:
rates_provider = dav2004r.rates_provider(table_type='AGGREGATE', estimate_type="BE")

To query for rates three risk factors need to be provided: CalendarYear, Gender and Age. Two of them are encoded in the canonical way, to be sure we get the gender right we import the corresponding *RiskFactor class* from pyprotolinc and then proceed creating three arrays with two entries each:

In [5]:
import numpy as np
from pyprotolinc.models.risk_factors import Gender

ages = np.array([60, 70], dtype=np.int16)
genders = np.array([Gender.M, Gender.F], dtype=np.int16)
calendar_years = np.array([2022, 2022], dtype=np.int16)

These arrays parametrize two individuals in the year 2022: a 60 year old male and a 70 year old female. For
both of them we can now lookup the mortality rates as follows.

In [6]:
rates_provider.get_rates(age=ages, gender=genders, calendaryear=calendar_years)

array([0.00392406, 0.00597981])

### Using the DAV Table via Configuration

The key change when compared to the example 1 is found in the *longevity_assumptions.yml* file which reads as follows:

In [7]:
with open('longevity_assumptions.yml', 'r') as f:
    print(f.read())

# for the annuity model
assumptions_spec:

  be:
    # annuitant's deaths
    - [0, 1, ["DAV2004R", "table_type:AGGREGATE", "estimate_type:BE",
              "base_directory:tables/Germany_Annuities_DAV2004R"]]

  res:
    # annuitant's deaths
    - [0, 1, ["DAV2004R", "table_type:AGGREGATE", "estimate_type:LOADED",
              "base_directory:tables/Germany_Annuities_DAV2004R"]]




Note that the parametrization is not a simple *scalar* as in example 1 but uses a syntax which should be
easy to grasp with the knowledge of the above.

We can now immediately start a run from the sheel by typeing ```pyprotolinc run``` in the main directory of this example. We can also start the run programmatically as follows.

In [12]:
from pyprotolinc.main import get_config_from_file, project_cashflows
run_config = get_config_from_file('config.yml')
project_cashflows(run_config);

INFO - 2022-07-03 12:29:11,827 - pyprotolinc.main - Multistate run with config: {'model_name': 'GenericMultiState', 'years_to_simulate': 119, 'portfolio_path': 'portfolio/portfolio_annuity_small.xlsx', 'assumptions_path': 'longevity_assumptions.yml', 'steps_per_month': 1, 'state_model_name': 'AnnuityRunoffStates', 'timestep_duration': 0.08333333333333333, 'outfile': 'results/ncf_out_generic.csv', 'portfolio_cache': 'portfolio/portfolio_cache', 'profile_out_dir': '.', 'portfolio_chunk_size': 1024, 'use_multicore': True}
INFO - 2022-07-03 12:29:11,891 - pyprotolinc.portfolio - Porfolio loaded from cache
INFO - 2022-07-03 12:29:11,893 - pyprotolinc.portfolio - Portolio rows: 1
DEBUG - 2022-07-03 12:29:11,895 - pyprotolinc.portfolio - Splitting portfolio for product ANNUITYINPAYMENT.
DEBUG - 2022-07-03 12:29:11,898 - pyprotolinc.portfolio - Initializing portfolio from dataframe
INFO - 2022-07-03 12:29:11,904 - pyprotolinc.main - Executions in single process for 1 units
INFO - 2022-07-03 12

### Rerun using Different Reserving Interest Rates
Let's have a quick look at the result and note that not only did we change the table but in this case also a reserving interest rate of 1% was used.

In [9]:
import pandas as pd
pd.read_csv("results/ncf_out_generic.csv", index_col=0).head()

YEAR  QUARTER  MONTH  PREMIUM  ANNUITY_PAYMENT1  ANNUITY_PAYMENT2  \
0  2021        4     12      0.0          0.000000               0.0   
1  2022        1      1      0.0       -100.000000               0.0   
2  2022        1      2      0.0        -99.988915               0.0   
3  2022        1      3      0.0        -99.977832               0.0   
4  2022        2      4      0.0        -99.966750               0.0   

   DEATH_PAYMENT  DI_LUMPSUM_PAYMENT  RESERVE_BOM(DIS1)  RESERVE_BOM(DEATH)  \
0            0.0                 0.0           0.000000                 0.0   
1            0.0                 0.0       45697.836566                 0.0   
2            0.0                 0.0       45634.456253                 0.0   
3            0.0                 0.0       45571.036132                 0.0   
4            0.0                 0.0       45507.576171                 0.0   

   ...  MV_ACTIVE_DIS1  MV_ACT_DIS2  MV_ACT_LAPSED  MV_ACT_MATURED  \
0  ...             0.0          0.0            0.0             0.0   
1  ...             0.0          0.0            0.0             0.0   
2  ...             0.0          0.0            0.0             0.0   
3  ...             0.0          0.0            0.0             0.0   
4  ...             0.0          0.0            0.0             0.0   

   MV_DIS1_DEATH  MV_DIS1_DIS2  MV_DIS1_ACT  MV_DIS2_DEATH  MV_DIS2_DIS1  \
0       0.000000           0.0          0.0            0.0           0.0   
1       0.000111           0.0          0.0            0.0           0.0   
2       0.000111           0.0          0.0            0.0           0.0   
3       0.000111           0.0          0.0            0.0           0.0   
4       0.000111           0.0          0.0            0.0           0.0   

   MV_DIS2_ACT  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 27 columns]

If we want to rerun using another reserve rate we can do that by loading and changing the portfolio as follows:

In [10]:
df_portfolio = pd.read_excel("portfolio/portfolio_annuity_small.xlsx")
df_portfolio.head()

DATE_PORTFOLIO  ID DATE_OF_BIRTH DATE_START_OF_COVER  SUM_INSURED  \
0     2021-12-31   1    1976-04-23          2022-01-01         1200   

  CURRENT_STATUS SEX           PRODUCT  PRODUCT_PARAMETERS SMOKERSTATUS  \
0           DIS1   m  AnnuityInPayment                 NaN            U   

   RESERVING_RATE  
0            0.01

In [11]:
df_portfolio["RESERVING_RATE"] = 0.03
res = project_cashflows(run_config, df_portfolio_overwrite=df_portfolio, export_to_file=False)
pd.DataFrame(res).head()

INFO - 2022-07-03 11:07:47,424 - pyprotolinc.main - Multistate run with config: {'model_name': 'GenericMultiState', 'years_to_simulate': 119, 'portfolio_path': 'portfolio/portfolio_annuity_small.xlsx', 'assumptions_path': 'longevity_assumptions.yml', 'steps_per_month': 1, 'state_model_name': 'AnnuityRunoffStates', 'timestep_duration': 0.08333333333333333, 'outfile': 'results/ncf_out_generic.csv', 'portfolio_cache': 'portfolio/portfolio_cache', 'profile_out_dir': '.', 'portfolio_chunk_size': 1024, 'use_multicore': True}
DEBUG - 2022-07-03 11:07:47,487 - pyprotolinc.portfolio - Initializing portfolio from dataframe
DEBUG - 2022-07-03 11:07:47,494 - pyprotolinc.portfolio - Splitting portfolio for product ANNUITYINPAYMENT.
DEBUG - 2022-07-03 11:07:47,497 - pyprotolinc.portfolio - Initializing portfolio from dataframe
INFO - 2022-07-03 11:07:47,502 - pyprotolinc.main - Executions in single process for 1 units
INFO - 2022-07-03 11:07:47,503 - pyprotolinc.main - Projecting subportfolio 1 / 1


YEAR  QUARTER  MONTH  PREMIUM  ANNUITY_PAYMENT1  ANNUITY_PAYMENT2  \
0  2021        4     12      0.0          0.000000               0.0   
1  2022        1      1      0.0       -100.000000               0.0   
2  2022        1      2      0.0        -99.988915               0.0   
3  2022        1      3      0.0        -99.977832               0.0   
4  2022        2      4      0.0        -99.966750               0.0   

   DEATH_PAYMENT  DI_LUMPSUM_PAYMENT  RESERVE_BOM(DIS1)  RESERVE_BOM(DEATH)  \
0            0.0                 0.0           0.000000                 0.0   
1            0.0                 0.0       30203.276681                 0.0   
2            0.0                 0.0       30176.722349                 0.0   
3            0.0                 0.0       30150.114342                 0.0   
4            0.0                 0.0       30123.452528                 0.0   

   ...  MV_ACTIVE_DIS1  MV_ACT_DIS2  MV_ACT_LAPSED  MV_ACT_MATURED  \
0  ...             0.0          0.0            0.0             0.0   
1  ...             0.0          0.0            0.0             0.0   
2  ...             0.0          0.0            0.0             0.0   
3  ...             0.0          0.0            0.0             0.0   
4  ...             0.0          0.0            0.0             0.0   

   MV_DIS1_DEATH  MV_DIS1_DIS2  MV_DIS1_ACT  MV_DIS2_DEATH  MV_DIS2_DIS1  \
0       0.000000           0.0          0.0            0.0           0.0   
1       0.000111           0.0          0.0            0.0           0.0   
2       0.000111           0.0          0.0            0.0           0.0   
3       0.000111           0.0          0.0            0.0           0.0   
4       0.000111           0.0          0.0            0.0           0.0   

   MV_DIS2_ACT  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 27 columns]

Note that the reserve is 30,203 when using 3% compared to 45,698 when using 1%.